File: src/services/authService.ts
import AsyncStorage from '@react-native-async-storage/async-storage';

// API Configuration - Using local server for development
// const API_BASE_URL = 'http://localhost:8000/api/v1';
const API_BASE_URL = 'https://dd41f2c65be9.ngrok-free.app/api/v1'; // ngrok URL (commented out)

// Storage Keys
const STORAGE_KEYS = {
  ACCESS_TOKEN: 'accessToken',
  REFRESH_TOKEN: 'refreshToken',
  USER_DATA: 'userData',
  IS_LOGGED_IN: 'isLoggedIn',
};

// Types
export interface LoginRequest {
  phone_number: string;
  otp: string;
}

export interface RegisterRequest {
  phone_number: string;
  email?: string;
  first_name?: string;
  last_name?: string;
  password?: string;
  referral_code?: string;
}

export interface OTPRequest {
  phone_number: string;
}

export interface AuthResponse {
  success: boolean;
  message: string;
  data?: {
    access_token: string;
    refresh_token: string;
    token_type: string;
    expires_in: number;
    user_id: string;
    role: string;
  };
}

export interface UserData {
  id: string;
  phone_number: string;
  email?: string;
  first_name?: string;
  last_name?: string;
  role: string;
  is_verified: boolean;
  total_points: number;
  created_at: string;
  last_login?: string;
}

// API Functions
const apiRequest = async (endpoint: string, options: RequestInit = {}) => {
  const url = `${API_BASE_URL}${endpoint}`;
  
  const config: RequestInit = {
    headers: {
      'Content-Type': 'application/json',
      ...options.headers,
    },
    ...options,
  };

  try {
    const response = await fetch(url, config);
    const data = await response.json();
    
    if (!response.ok) {
      throw new Error(data.detail || 'API request failed');
    }
    
    return data;
  } catch (error) {
    console.error('API request error:', error);
    throw error;
  }
};

const authenticatedRequest = async (endpoint: string, options: RequestInit = {}) => {
  const token = await getAccessToken();
  if (!token) {
    throw new Error('No access token available');
  }

  return apiRequest(endpoint, {
    ...options,
    headers: {
      ...options.headers,
      'Authorization': `Bearer ${token}`,
    },
  });
};

// Authentication Functions
export const sendOTP = async (phoneNumber: string): Promise<AuthResponse> => {
  return apiRequest('/auth/send-otp', {
    method: 'POST',
    body: JSON.stringify({ phone_number: phoneNumber }),
  });
};

export const verifyOTP = async (phoneNumber: string, otp: string): Promise<AuthResponse> => {
  return apiRequest('/auth/verify-otp', {
    method: 'POST',
    body: JSON.stringify({ phone_number: phoneNumber, otp }),
  });
};

export const login = async (phoneNumber: string, otp: string): Promise<AuthResponse> => {
  return apiRequest('/auth/login', {
    method: 'POST',
    body: JSON.stringify({ phone_number: phoneNumber, otp }),
  });
};

export const register = async (userData: RegisterRequest): Promise<AuthResponse> => {
  return apiRequest('/auth/register', {
    method: 'POST',
    body: JSON.stringify(userData),
  });
};

export const refreshToken = async (): Promise<AuthResponse> => {
  const refreshToken = await getRefreshToken();
  if (!refreshToken) {
    throw new Error('No refresh token available');
  }

  return apiRequest('/auth/refresh', {
    method: 'POST',
    body: JSON.stringify({ refresh_token: refreshToken }),
  });
};

export const logout = async (): Promise<void> => {
  try {
    const refreshToken = await getRefreshToken();
    const accessToken = await getAccessToken();
    
    if (refreshToken && accessToken) {
      await apiRequest('/auth/logout', {
        method: 'POST',
        body: JSON.stringify({ refresh_token: refreshToken }),
        headers: {
          'Authorization': `Bearer ${accessToken}`,
        },
      });
    }
  } catch (error) {
    console.error('Logout error:', error);
  } finally {
    await clearAllTokens();
  }
};

export const getCurrentUser = async (): Promise<UserData> => {
  const response = await authenticatedRequest('/auth/me');
  return response.data;
};

// Token Management
export const storeTokens = async (accessToken: string, refreshToken: string) => {
  try {
    await AsyncStorage.multiSet([
      [STORAGE_KEYS.ACCESS_TOKEN, accessToken],
      [STORAGE_KEYS.REFRESH_TOKEN, refreshToken],
      [STORAGE_KEYS.IS_LOGGED_IN, 'true'],
    ]);
  } catch (error) {
    console.error('Error storing tokens:', error);
  }
};

export const getAccessToken = async (): Promise<string | null> => {
  try {
    const token = await AsyncStorage.getItem(STORAGE_KEYS.ACCESS_TOKEN);
    console.log('AuthService: getAccessToken result:', { hasToken: !!token });
    return token;
  } catch (error) {
    console.error('AuthService: Error getting access token:', error);
    return null;
  }
};

export const getRefreshToken = async (): Promise<string | null> => {
  try {
    const token = await AsyncStorage.getItem(STORAGE_KEYS.REFRESH_TOKEN);
    console.log('AuthService: getRefreshToken result:', { hasToken: !!token });
    return token;
  } catch (error) {
    console.error('AuthService: Error getting refresh token:', error);
    return null;
  }
};

export const clearAllTokens = async () => {
  try {
    await AsyncStorage.multiRemove([
      STORAGE_KEYS.ACCESS_TOKEN,
      STORAGE_KEYS.REFRESH_TOKEN,
      STORAGE_KEYS.USER_DATA,
      STORAGE_KEYS.IS_LOGGED_IN,
    ]);
  } catch (error) {
    console.error('Error clearing tokens:', error);
  }
};

export const isLoggedIn = async (): Promise<boolean> => {
  try {
    const token = await getAccessToken();
    const loginStatus = await AsyncStorage.getItem(STORAGE_KEYS.IS_LOGGED_IN);
    const result = !!(token && loginStatus === 'true');
    console.log('AuthService: isLoggedIn check:', { hasToken: !!token, loginStatus, result });
    return result;
  } catch (error) {
    console.error('AuthService: Error checking login status:', error);
    return false;
  }
};

// User Data Management
export const storeUserData = async (userData: UserData) => {
  try {
    await AsyncStorage.setItem(STORAGE_KEYS.USER_DATA, JSON.stringify(userData));
  } catch (error) {
    console.error('Error storing user data:', error);
  }
};

export const getUserData = async (): Promise<UserData | null> => {
  try {
    const userData = await AsyncStorage.getItem(STORAGE_KEYS.USER_DATA);
    const result = userData ? JSON.parse(userData) : null;
    console.log('AuthService: getUserData result:', { hasUserData: !!result });
    return result;
  } catch (error) {
    console.error('AuthService: Error getting user data:', error);
    return null;
  }
};

// Token Refresh Logic
export const refreshAccessToken = async (): Promise<string | null> => {
  try {
    const response = await refreshToken();
    if (response.success && response.data) {
      await storeTokens(response.data.access_token, response.data.refresh_token);
      return response.data.access_token;
    }
    return null;
  } catch (error) {
    console.error('Error refreshing token:', error);
    await clearAllTokens();
    return null;
  }
};

// Auto-refresh token when making authenticated requests
export const authenticatedRequestWithRefresh = async (endpoint: string, options: RequestInit = {}) => {
  try {
    return await authenticatedRequest(endpoint, options);
  } catch (error: any) {
    if (error.message?.includes('401') || error.message?.includes('token')) {
      // Token expired, try to refresh
      const newToken = await refreshAccessToken();
      if (newToken) {
        // Retry with new token
        return await authenticatedRequest(endpoint, options);
      }
    }
    throw error;
  }
}; 

File: app/services/auth_service.py
from datetime import datetime, timedelta
from typing import Optional, Dict, Any
from sqlalchemy.orm import Session
from app.core.database import get_db
from app.core.security import (
    create_access_token, create_refresh_token, verify_token,
    generate_otp, generate_user_id
)
from app.models.user import UserCreate, UserInDB, UserResponse
from app.schemas.auth import TokenResponse, RegisterRequest
from app.core.config import settings
import logging
import redis
try:
    from twilio.rest import Client
    from twilio.base.exceptions import TwilioException
    TWILIO_AVAILABLE = True
except ImportError:
    TWILIO_AVAILABLE = False

logger = logging.getLogger(__name__)

class AuthService:
    def __init__(self, db: Session):
        self.db = db
        self.redis_client = redis.from_url(settings.REDIS_URL, decode_responses=True)
    
    async def send_otp(self, phone_number: str) -> Dict[str, Any]:
        """Send OTP to phone number."""
        try:
            # Generate OTP
            otp = generate_otp()
            
            # Store OTP in Redis with expiration (5 minutes)
            otp_key = f"otp:{phone_number}"
            self.redis_client.setex(otp_key, 300, otp)
            
            # Send OTP via Twilio SMS
            if TWILIO_AVAILABLE and settings.TWILIO_ACCOUNT_SID and settings.TWILIO_AUTH_TOKEN:
                try:
                    client = Client(settings.TWILIO_ACCOUNT_SID, settings.TWILIO_AUTH_TOKEN)
                    message = client.messages.create(
                        body=f"Your Trison Solar OTP is: {otp}. Valid for 5 minutes.",
                        from_=settings.TWILIO_PHONE_NUMBER,
                        to=f"+{phone_number}" if not phone_number.startswith('+') else phone_number
                    )
                    logger.info(f"OTP SMS sent successfully to {phone_number}. SID: {message.sid}")
                except TwilioException as twilio_error:
                    logger.error(f"Twilio SMS error: {twilio_error}")
                    # Fallback to logging OTP if SMS fails
                    logger.info(f"OTP for {phone_number}: {otp}")
            else:
                # Fallback: log OTP if Twilio not configured
                logger.info(f"OTP for {phone_number}: {otp}")
                logger.warning("Twilio not configured. OTP only logged to console.")
                
                # In development mode, return OTP in response for testing
                if settings.DEBUG:
                    return {"success": True, "otp": otp, "message": "OTP generated (Twilio not configured)"}
                
                return {"success": True, "message": "OTP sent successfully"}
        except Exception as e:
            logger.error(f"Error sending OTP: {e}")
            raise e
    
    async def verify_otp(self, phone_number: str, otp: str) -> TokenResponse:
        """Verify OTP and return tokens."""
        try:
            # Get stored OTP from Redis
            otp_key = f"otp:{phone_number}"
            stored_otp = self.redis_client.get(otp_key)
            
            if not stored_otp or stored_otp != otp:
                raise ValueError("Invalid OTP")
            
            # Delete OTP from Redis after successful verification
            self.redis_client.delete(otp_key)
            
            # Check if user exists in PostgreSQL
            from app.models.user import User
            user = self.db.query(User).filter(User.phone_number == phone_number).first()
            
            if not user:
                # Create new user if doesn't exist
                user_data = UserCreate(
                    phone_number=phone_number,
                    is_verified=True
                )
                user = await self._create_user(user_data)
            else:
                # Update user verification status
                user.is_verified = True
                user.last_login = datetime.utcnow()
                user.login_count = user.login_count + 1 if user.login_count else 1
                user.updated_at = datetime.utcnow()
                self.db.commit()
            
            # Generate tokens
            access_token = create_access_token(data={"sub": str(user.id), "role": user.role})
            refresh_token = create_refresh_token(data={"sub": str(user.id)})
            
            # Log token information for debugging
            logger.info(f"=== USER LOGIN SUCCESSFUL ===")
            logger.info(f"User ID: {user.id}")
            logger.info(f"Phone: {user.phone_number}")
            logger.info(f"Name: {user.first_name} {user.last_name}")
            logger.info(f"Role: {user.role}")
            logger.info(f"Points: {user.total_points}")
            logger.info(f"Access Token: {access_token[:50]}...")
            logger.info(f"Refresh Token: {refresh_token[:50]}...")
            logger.info(f"Token Expires: {settings.ACCESS_TOKEN_EXPIRE_MINUTES} minutes")
            logger.info(f"================================")
            
            return TokenResponse(
                access_token=access_token,
                refresh_token=refresh_token,
                token_type="bearer",
                expires_in=settings.ACCESS_TOKEN_EXPIRE_MINUTES * 60,
                user_id=str(user.id),
                role=user.role
            )
            
        except Exception as e:
            logger.error(f"Error verifying OTP: {e}")
            raise e
    
    async def login(self, phone_number: str, otp: str) -> TokenResponse:
        """Login user with phone number and OTP."""
        return await self.verify_otp(phone_number, otp)
    
    async def register(self, request: RegisterRequest) -> TokenResponse:
        """Register new user."""
        try:
            # Check if user already exists
            from app.models.user import User
            existing_user = self.db.query(User).filter(User.phone_number == request.phone_number).first()
            if existing_user:
                raise ValueError("User already exists")
            
            # Create user data
            user_data = UserCreate(
                phone_number=request.phone_number,
                email=request.email,
                first_name=request.first_name,
                last_name=request.last_name,
                is_verified=True
            )
            
            # Create user
            user = await self._create_user(user_data)
            
            # Generate tokens
            access_token = create_access_token(data={"sub": str(user.id), "role": user.role})
            refresh_token = create_refresh_token(data={"sub": str(user.id)})
            
            return TokenResponse(
                access_token=access_token,
                refresh_token=refresh_token,
                token_type="bearer",
                expires_in=settings.ACCESS_TOKEN_EXPIRE_MINUTES * 60,
                user_id=str(user.id),
                role=user.role
            )
            
        except Exception as e:
            logger.error(f"Error during registration: {e}")
            raise e
    
    async def refresh_token(self, refresh_token: str) -> TokenResponse:
        """Refresh access token using refresh token."""
        try:
            # Verify refresh token
            payload = verify_token(refresh_token)
            if not payload or payload.get("type") != "refresh":
                raise ValueError("Invalid refresh token")
            
            user_id = payload.get("sub")
            from app.models.user import User
            user = self.db.query(User).filter(User.id == int(user_id)).first()
            
            if not user:
                raise ValueError("User not found")
            
            # Generate new tokens
            access_token = create_access_token(data={"sub": user_id, "role": user.role})
            new_refresh_token = create_refresh_token(data={"sub": user_id})
            
            return TokenResponse(
                access_token=access_token,
                refresh_token=new_refresh_token,
                token_type="bearer",
                expires_in=settings.ACCESS_TOKEN_EXPIRE_MINUTES * 60,
                user_id=user_id,
                role=user.role
            )
            
        except Exception as e:
            logger.error(f"Error refreshing token: {e}")
            raise e
    
    async def logout(self, refresh_token: str, access_token: str) -> bool:
        """Logout user and invalidate tokens."""
        try:
            # Add tokens to blacklist
            self.redis_client.setex(f"blacklist:{access_token}", 3600, "1")  # 1 hour
            self.redis_client.setex(f"blacklist:{refresh_token}", 86400, "1")  # 24 hours
            
            return True
        except Exception as e:
            logger.error(f"Error during logout: {e}")
            raise e
    
    async def get_current_user(self, user_id: str) -> Dict[str, Any]:
        """Get current user information."""
        try:
            from app.models.user import User
            user = self.db.query(User).filter(User.id == int(user_id)).first()
            if not user:
                raise ValueError("User not found")
            
            return {
                "id": str(user.id),
                "phone_number": user.phone_number,
                "email": user.email,
                "first_name": user.first_name,
                "last_name": user.last_name,
                "role": user.role,
                "is_verified": user.is_verified,
                "total_points": user.total_points,
                "created_at": user.created_at,
                "last_login": user.last_login
            }
        except Exception as e:
            logger.error(f"Error getting current user: {e}")
            raise e
    
    async def _create_user(self, user_data: UserCreate) -> Any:
        """Create new user in database."""
        try:
            from app.models.user import User
            
            user_dict = user_data.dict()
            user_dict["created_at"] = datetime.utcnow()
            user_dict["updated_at"] = datetime.utcnow()
            user_dict["last_login"] = datetime.utcnow()
            user_dict["login_count"] = 1
            
            # No password needed for OTP-based authentication
            # user_dict["hashed_password"] = None
            
            # Generate referral code
            user_dict["referral_code"] = f"TRS{generate_user_id()[:8].upper()}"
            
            # Create user instance
            user = User(**user_dict)
            self.db.add(user)
            self.db.commit()
            self.db.refresh(user)
            
            return user
        except Exception as e:
            logger.error(f"Error creating user: {e}")
            self.db.rollback()
            raise e 